apollo-router/src/plugins/authorization/mod.rs

```diff
@@ -1,31 +1,323 @@
+    experimental_enable_authorization_directives: bool,
 }
 
-struct AuthorizationPlugin {
-    enabled: bool,
+pub(crate) struct AuthorizationPlugin {
+    require_authentication: bool,
+}
+
+impl AuthorizationPlugin {
+    pub(crate) fn enable_directives(
+        configuration: &Configuration,
+        schema: &Schema,
+    ) -> Result<bool, ServiceBuildError> {
+        let has_config = configuration
+            .apollo_plugins
+            .plugins
+            .iter()
+            .find(|(s, _)| s.as_str() == "authorization")
+            .and_then(|(_, v)| {
+                v.get("experimental_enable_authorization_directives")
+                    .and_then(|v| v.as_bool())
+            });
+        let has_authorization_directives = schema
+            .type_system
+            .definitions
+            .directives
+            .contains_key(AUTHENTICATED_DIRECTIVE_NAME)
+            || schema
+                .type_system
+                .definitions
+                .directives
+                .contains_key(REQUIRES_SCOPES_DIRECTIVE_NAME);
+
+        match has_config {
+            Some(b) => Ok(b),
+            None => {
+                if has_authorization_directives {
+                    Err(ServiceBuildError::Schema(SchemaError::Api("cannot start the router on a schema with authorization directives without configuring the authorization plugin".to_string())))
+                } else {
+                    Ok(false)
+                }
+            }
+        }
+    }
+
+    pub(crate) async fn query_analysis(
+        query: &str,
+        schema: &Schema,
+        configuration: &Configuration,
+        context: &Context,
+    ) {
+        let (compiler, file_id) = Query::make_compiler(query, schema, configuration);
+
+        let mut visitor = ScopeExtractionVisitor::new(&compiler, file_id);
+
+        // if this fails, the query is invalid and will fail at the query planning phase.
+        // We do not return validation errors here for now because that would imply a huge
+        // refactoring of telemetry and tests
+        if traverse::document(&mut visitor, file_id).is_ok() {
+            let scopes: Vec<String> = visitor.extracted_scopes.into_iter().collect();
+
+            context.insert(REQUIRED_SCOPES_KEY, scopes).unwrap();
+        }
+    }
+
+    pub(crate) fn update_cache_key(context: &Context) {
+        let is_authenticated = context.contains_key(APOLLO_AUTHENTICATION_JWT_CLAIMS);
+
+        let request_scopes = context
+            .get_json_value(APOLLO_AUTHENTICATION_JWT_CLAIMS)
+            .and_then(|value| {
+                value.as_object().and_then(|object| {
+                    object.get("scope").and_then(|v| {
+                        v.as_str()
+                            .map(|s| s.split(' ').map(|s| s.to_string()).collect::<HashSet<_>>())
+                    })
+                })
+            });
+        let query_scopes = context.get_json_value(REQUIRED_SCOPES_KEY).and_then(|v| {
```

Comment:
this (and other new entries) should probably go in private entries instead

Reply:
this one cannot go in private entries because it is cached by query analysis, and private entries cannot be cloned